### DISCUSSION TITLE
Is there a clean .svs[level=1] to .tiff file conversion

### DISCUSSION THREAD

**Discussion by VincePlantItAi** [libvips/libvips#2370] [Q&A]
Hello there,
here our lab produces .svs files that are too large, with unncessarily high resolution at level 0
So I have been trying to remove level 0 and only take images from level 1,2,3, and so on, without corrupting images.

But no success so far for about a month or two.
Since svs-writer does not exist, I decided to utilize "vips" to generate tiff file from level 1.

Even after trying almost all possible combination of functions & options,
if we view at viewrs like ASAP or DeepZoom_server.py,
All resulting .tiff files are slightly less clear than the original .svs file at level resolution 
The hospital wants the output tiff file to contain EXACTLY same images as the original svs file (from level 1 and on)

here is original file, the share link will expire in 48hours, lemme know if you need more time mr.jcupitt !
https://planit20-my.sharepoint.com/:u:/g/personal/vince_planit_ai/EXWbudNPz8FKhX-06L55TMMBqptrmCKEolpKAUN3Q6YBzQ?e=EhkQRB

( check below screenshot for comparison for couple of results ).
![Screen Shot 2021-07-29 at 3 26 31 PM](https://user-images.githubusercontent.com/72775544/127487465-7455438b-a683-4b66-ab6e-3159c421e9a5.png)
![Screen Shot 2021-07-29 at 3 27 31 PM](https://user-images.githubusercontent.com/72775544/127487491-e43a6682-9654-4753-b30a-6410c4359f00.png)
![Screen Shot 2021-07-29 at 3 28 42 PM](https://user-images.githubusercontent.com/72775544/127487500-335a956d-f266-4efe-b375-a4e6d407561e.png)
![Screen Shot 2021-07-29 at 3 29 53 PM](https://user-images.githubusercontent.com/72775544/127487507-8d65638a-098a-409f-9cd2-90ed12d4d98a.png)
![Screen Shot 2021-07-29 at 3 52 07 PM](https://user-images.githubusercontent.com/72775544/127487511-1c29ea38-28ee-4148-90cf-dd93a03ab566.png)

below are some of the scripts that I have tried
(MACOS)
`vips copy slide_x400.svs\[level=1\] slide_x400/slide_x400-jcu-bands4.tiff\[pyramid,tile,bigtiff,compression=jpeg,Q=90,tile-width=256,tile-height=256,resunit=inch\] --bands 4`
`vips openslideload slide_x400.svs slide_x400-df-04.tiff [tile,pyramid,properties,lossless,compression=deflate,resunit=inch,Q=100\] --level 1`
`vips extract_band CMU-1.svs CMU-0-3-1.tiff\[pyramid,tile,Q=30,compression=jpeg,tile-width=256,tile-height=256\] 0 --n 3`
`vips extract_band CMU-1.svs CMU-0-3-2.tiff\[pyramid,tile,Q=30,tile-width=256,tile-height=256\] 0 --n 3`

**Comment by jcupitt** [comment]
Hi, your sharepoint link needs authentication and won't open for me. Maybe try google drive or dropbox?

**Comment by jcupitt** [comment]
libvips will always decompress and recompress, so there will be very small differences with JPEG tiles. If you use a lossless compressor (like deflate), then there will be no difference.

What differences are the hospital seeing exactly? Could they be caused by the viewer software?

**Comment by JooHyukKim** [comment]
Thank you for your kind reply

A ) here is url to original file https://www.dropbox.com/s/p9pjl3hgp8xvk0n/slide_x400.svs?dl=0

> libvips will always decompress and recompress, so there will be very small differences with JPEG tiles. If you use a lossless compressor (like deflate), then there will be no difference.

Q ) I see, so straight vips copy/tiffsave/openslideload with lossless compressors like deflate would result in conversion with no difference made to the images then?

Q ) is deflate the only lossless compressor option, or is there any other?

> What differences are the hospital seeing exactly?

They see the output .tiff level[0] of slide image slightly less clear than than level [1] of original svs file 😢 

> Could they be caused by the viewer software?

Might be, the hopsital and I are both using both deepzoom_server.py & ASAP(it's a name)

Deepzoom_server.py : https://openslide.org/api/python/
ASAP : https://computationalpathologygroup.github.io/ASAP/

cheers.

**Comment by jcupitt** [comment]
You could try JPEG Q=100, the differences should be extremely small.

Another alternative would be to just remove the highest res layer after conversion. That really should ensure they are identical. 

For example:

```
john@kiwi:~/pics$ vips dzsave k2.jpg x
john@kiwi:~/pics$ cd x_files/
john@kiwi:~/pics/x_files$ ls
0  1  10  11  2  3  4  5  6  7  8  9
john@kiwi:~/pics/x_files$ rm -rf 11
```

Then adjust `x.dzi`:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Image xmlns="http://schemas.microsoft.com/deepzoom/2008"
  Format="jpeg"
  Overlap="1"
  TileSize="254"
  >
  <Size 
    Height="2048"
    Width="1450"
  />
</Image>
```

Just calculate `(Height + 1) / 2`, ie. half with round up. It'd be simple to script with bash.

**Comment by jcupitt** [comment]
Sorry, I've just realized that you're not converting the slides, you're using `deepzoom_server.py`.

You could try conversion to DZI and serving that instead. Set Q high in the conversion and you should be fine. It'll be much faster for viewing.

If you must use the py server, check the code for tile conversion to JPEG and look at the Q factor there. You'll want to disable chroma subsample.
